File: backend/app/schemas/timeline.py
"""Pydantic schemas for the timeline-graph domain (R6-2 时间线图谱).

The timeline view exposes the novel's plot events as a causal DAG: nodes
(events) ordered by in-world date / chapter index, edges following
prev_event_id predecessor pointers, plus any structural warnings (dangling
predecessor, reverse ordering, cycles). `topological_order` is the causal
topological sort of event ids.
"""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class TimelineNode(BaseModel):
    """A plot event as a node in the causal DAG."""

    event_id: int
    event_type: str = Field(default="beat")
    summary: str
    chapter_id: int | None = None
    chapter_index: int | None = None
    in_world_date: str | None = None
    prev_event_id: int | None = None


class TimelineEdge(BaseModel):
    """Directed edge `from_id` (predecessor) -> `to_id` (successor)."""

    from_id: int
    to_id: int


class TimelineWarning(BaseModel):
    """A timeline-consistency finding (real-time conflict warning)."""

    model_config = ConfigDict(frozen=True)

    kind: str = Field(..., description="predecessor|reverse_order|cycle")
    event_id: int
    detail: str


class TimelineResponse(BaseModel):
    """Full timeline view for one novel."""

    nodes: list[TimelineNode]
    edges: list[TimelineEdge]
    warnings: list[TimelineWarning]
    topological_order: list[int]

File: frontend/src/lib/timeline.ts
/**
 * Timeline API client — causal DAG of plot events (R6-2).
 * Endpoint (backend app/api/timeline.py):
 *   GET /v1/documents/{docId}/timeline
 */
import { ApiError } from "@/lib/types";
import { backendUrl } from "@/lib/config";
import { ownerAuthHeaders } from "@/lib/settings";

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${backendUrl}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...ownerAuthHeaders(),
      ...(init?.headers ?? {}),
    },
  });
  if (!res.ok) {
    let detail: unknown;
    try {
      const body = await res.json();
      detail = body?.detail ?? body;
    } catch {
      // non-JSON error body
    }
    const msg =
      typeof detail === "string" ? detail : `请求失败 (${res.status})`;
    throw new ApiError(msg, res.status, detail);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

export interface TimelineNode {
  event_id: number;
  event_type: string;
  summary: string;
  chapter_id: number | null;
  chapter_index: number | null;
  in_world_date: string | null;
}

export interface TimelineEdge {
  from_id: number;
  to_id: number;
}

export interface TimelineWarning {
  kind: "predecessor" | "reverse_order" | "cycle";
  event_id: number;
  detail: string;
}

export interface TimelineResponse {
  nodes: TimelineNode[];
  edges: TimelineEdge[];
  warnings: TimelineWarning[];
  topological_order: number[];
}

export async function fetchTimeline(
  docId: number,
): Promise<TimelineResponse> {
  return request<TimelineResponse>(`/v1/documents/${docId}/timeline`);
}
